File: sol/502/502.py
import heapq as hq

class Solution:
    def findMaximizedCapital(self, k: int, w: int, profits: List[int], capital: List[int]) -> int:
        n = len(profits)
        indices = sorted(range(n), key=lambda i: capital[i])
        ptr = 0
        h = []
        for _ in range(k):
            while ptr < n and capital[indices[ptr]] <= w:
                hq.heappush(h, -profits[indices[ptr]])
                ptr += 1
            if len(h) == 0:
                break
            w -= hq.heappop(h)
        return w

File: sol/502/502.cpp
struct Solution {
    struct Proj {
        int profit;
        int capital;

        friend bool operator<(const Proj &a, const Proj &b) {
            return a.profit < b.profit;
        }
    };
    int findMaximizedCapital(int k, int w, vector<int>& profits, vector<int>& capital) {
        int n = profits.size(), ptr = 0;
        vector<Proj> projs(n), h;
        for (int i = 0; i < n; ++i){
            projs[i] = Proj(profits[i], capital[i]);
        }
        sort(projs.begin(), projs.end(), [](const Proj &a, const Proj &b){return a.capital < b.capital;});
        while (k-- > 0){
            while (ptr < n && projs[ptr].capital <= w){
                h.push_back(projs[ptr++]);
                push_heap(h.begin(), h.end());
            }
            if (h.empty()){
                break;
            }
            pop_heap(h.begin(), h.end());
            w += h.back().profit;
            h.pop_back();
        }
        return w;
    }
};
